File: csharp/AssemblerWeb/AppSitesConfig.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AssemblerWeb
{
    public static class AppSitesConfig
    {
        /// <summary>
        /// Discovers AppSites by scanning the AppSites folder and generates appsites.csv
        /// </summary>
        private static void GenerateAppSitesCsv(string wwwrootPath)
        {
            var appSitesPath = Path.Combine(wwwrootPath, "AppSites");
            var appDataPath = Path.Combine(wwwrootPath, "App_Data");
            var csvFilePath = Path.Combine(appDataPath, "appsites.csv");

            if (!Directory.Exists(appSitesPath))
            {
                throw new DirectoryNotFoundException($"AppSites directory not found: {appSitesPath}");
            }

            // Ensure App_Data directory exists
            if (!Directory.Exists(appDataPath))
            {
                Directory.CreateDirectory(appDataPath);
            }

            // Get all directories in AppSites folder
            var appSites = Directory.GetDirectories(appSitesPath)
                .Select(dir => Path.GetFileName(dir))
                .Where(dirName => !string.IsNullOrEmpty(dirName))
                .OrderBy(name => name)
                .ToList();

            // Add Index as it's a valid AppSite
            if (!appSites.Contains("Index"))
            {
                appSites.Add("Index");
            }

            // Write as CSV (comma-delimited)
            var csv = string.Join(",", appSites);
            File.WriteAllText(csvFilePath, csv);

            Console.WriteLine($"[AppSitesConfig] Generated appsites.csv with {appSites.Count} AppSites");
        }

        /// <summary>
        /// Loads AppSites from appsites.csv, generates it if it doesn't exist
        /// </summary>
        public static HashSet<string> LoadAppSites(string wwwrootPath)
        {
            var appDataPath = Path.Combine(wwwrootPath, "App_Data");
            var csvFilePath = Path.Combine(appDataPath, "appsites.csv");

            // Generate appsites.csv if it doesn't exist
            if (!File.Exists(csvFilePath))
            {
                Console.WriteLine($"[AppSitesConfig] appsites.csv not found, generating...");
                GenerateAppSitesCsv(wwwrootPath);
            }

            // Read and parse CSV
            var csv = File.ReadAllText(csvFilePath).Trim();

            if (string.IsNullOrWhiteSpace(csv))
            {
                throw new InvalidOperationException("appsites.csv is empty");
            }

            var appSites = csv.Split(',')
                .Select(s => s.Trim())
                .Where(s => !string.IsNullOrWhiteSpace(s))
                .ToList();

            if (appSites.Count == 0)
            {
                throw new InvalidOperationException("No AppSites found in appsites.csv");
            }

            Console.WriteLine($"[AppSitesConfig] Loaded {appSites.Count} AppSites from appsites.csv");

            return new HashSet<string>(appSites, StringComparer.OrdinalIgnoreCase);
        }

        /// <summary>
        /// Reloads AppSites by regenerating appsites.csv from the file system
        /// </summary>
        public static HashSet<string> ReloadAppSites(string wwwrootPath)
        {
            Console.WriteLine($"[AppSitesConfig] Reloading AppSites...");
            GenerateAppSitesCsv(wwwrootPath);
            return LoadAppSites(wwwrootPath);
        }
    }
}

File: go/AssemblerWeb/app_sites_config.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// GenerateAppSitesCsv discovers AppSites by scanning the AppSites folder and generates appsites.csv
func GenerateAppSitesCsv(wwwrootPath string) error {
	appSitesPath := filepath.Join(wwwrootPath, "AppSites")
	appDataPath := filepath.Join(wwwrootPath, "App_Data")
	csvFilePath := filepath.Join(appDataPath, "appsites.csv")

	if _, err := os.Stat(appSitesPath); os.IsNotExist(err) {
		return fmt.Errorf("AppSites directory not found: %s", appSitesPath)
	}

	// Ensure App_Data directory exists
	if err := os.MkdirAll(appDataPath, 0755); err != nil {
		return fmt.Errorf("failed to create App_Data directory: %w", err)
	}

	// Get all directories in AppSites folder
	entries, err := os.ReadDir(appSitesPath)
	if err != nil {
		return fmt.Errorf("failed to read AppSites directory: %w", err)
	}

	var appSites []string
	for _, entry := range entries {
		if entry.IsDir() {
			appSites = append(appSites, entry.Name())
		}
	}

	// Add Index as it's a valid AppSite
	hasIndex := false
	for _, site := range appSites {
		if strings.EqualFold(site, "Index") {
			hasIndex = true
			break
		}
	}
	if !hasIndex {
		appSites = append(appSites, "Index")
	}

	// Sort for consistency
	sort.Strings(appSites)

	// Write as CSV (comma-delimited)
	csv := strings.Join(appSites, ",")
	if err := os.WriteFile(csvFilePath, []byte(csv), 0644); err != nil {
		return fmt.Errorf("failed to write appsites.csv: %w", err)
	}

	fmt.Printf("[AppSitesConfig] Generated appsites.csv with %d AppSites\n", len(appSites))
	return nil
}

// LoadAppSites loads AppSites from appsites.csv, generates it if it doesn't exist
func LoadAppSites(wwwrootPath string) (map[string]bool, error) {
	appDataPath := filepath.Join(wwwrootPath, "App_Data")
	csvFilePath := filepath.Join(appDataPath, "appsites.csv")

	// Generate appsites.csv if it doesn't exist
	if _, err := os.Stat(csvFilePath); os.IsNotExist(err) {
		fmt.Println("[AppSitesConfig] appsites.csv not found, generating...")
		if err := GenerateAppSitesCsv(wwwrootPath); err != nil {
			return nil, err
		}
	}

	// Read and parse CSV
	data, err := os.ReadFile(csvFilePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read appsites.csv: %w", err)
	}

	csv := strings.TrimSpace(string(data))
	if csv == "" {
		return nil, fmt.Errorf("appsites.csv is empty")
	}

	parts := strings.Split(csv, ",")
	appSites := make(map[string]bool)
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed != "" {
			appSites[trimmed] = true
		}
	}

	if len(appSites) == 0 {
		return nil, fmt.Errorf("no AppSites found in appsites.csv")
	}

	fmt.Printf("[AppSitesConfig] Loaded %d AppSites from appsites.csv\n", len(appSites))
	return appSites, nil
}

// ReloadAppSites reloads AppSites by regenerating appsites.csv from the file system
func ReloadAppSites(wwwrootPath string) (map[string]bool, error) {
	fmt.Println("[AppSitesConfig] Reloading AppSites...")
	if err := GenerateAppSitesCsv(wwwrootPath); err != nil {
		return nil, err
	}
	return LoadAppSites(wwwrootPath)
}
